### ISSUE TITLE
[Question]: Algorithm behind low communication cost for matrix multiplication in Semi2k

### ISSUE THREAD

**Issue by winnylyc** [secretflow/spu#590]
### Issue Type

Performance

### Modules Involved

MPC protocol

### Have you reproduced the bug with SPU HEAD?

Yes

### Have you searched existing issues?

Yes

### SPU Version

spu 0.7.0b0

### OS Platform and Distribution

Linux Ubuntu 22.04

### Python Version

3.10

### Current Behavior?

Hi, SPU team! Sorry for disturbing you.

I am doing some simple benchmark for SPU. When I test the matrix multiplication, I found the "send bytes" for matrix multiplication in semi2k is much lower than I had expected.
Initially, I conducted a test with a 1x1 matrix multiplied by a 1x1 matrix, and the "send bytes" value was 16. Subsequently, I performed a test with a 4x4 matrix multiplied by a 4x4 matrix, and the "send bytes" value was 256. Based on this data, it appears that there are only 16 element-wise multiplications occurring, which is significantly lower than the expected 64 multiplications. 
I am curious to understand the optimization behind this performance. Is there any documents or papers about this?

Looking forward to your reply! Thanks!

### Standalone code to reproduce the issue

```Python
import secretflow as sf
import spu
import jax.numpy as jnp
import numpy as np

def test_function(x, y):
    result = jnp.matmul(x, y)
    return result

sf.shutdown()
sf.init(['alice', 'bob'], address='local')

alice, bob = sf.PYU('alice'), sf.PYU('bob')

semi2k_config = sf.utils.testing.cluster_def(
    parties=['alice', 'bob'],
    runtime_config={
        'protocol': spu.spu_pb2.SEMI2K,
        'field': spu.spu_pb2.FM64,
        "enable_pphlo_profile": True,
        "enable_hal_profile": True,
    },
)

spu_device = sf.SPU(semi2k_config)

x_plain = np.array([[5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.]])
# x_plain =  np.array([5.])

y_plain = np.array([[5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.]])
# y_plain = np.array([5.])

x = sf.to(alice, x_plain)
y = sf.to(bob, y_plain)

result = spu_device(test_function)(x, y)

print(sf.reveal(result))
```

### Relevant log output

```Shell
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.748 [info] [api.cc:printProfilingData:158] [Profiling] SPU execution test_function completed, input processing took 1.6e-06s, execution took 0.001509273s, output processing took 1.085e-06s, total time 0.001511958s.
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:191] HLO profiling: total time 0.000780761
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:194] - pphlo.dot, executed 1 times, duration 0.000780761s, send bytes 384
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:191] HAL profiling: total time 0.000755631
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:194] - f_mmul, executed 1 times, duration 0.000755631s, send bytes 384
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:191] MPC profiling: total time 0.000739545
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:194] - mmul_aa, executed 1 times, duration 0.000695586s, send bytes 256
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:194] - reshape, executed 1 times, duration 3.788e-06s, send bytes 0
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:194] - trunc_a, executed 1 times, duration 4.0171e-05s, send bytes 128
(SPURuntime(device_id=None, party=bob) pid=97554) 2024-03-07 12:03:06.749 [info] [api.cc:printProfilingData:204] Link details: total send bytes 384, send actions 2
```

**Comment by tpppppub** [comment]
you can check the semi2k [mmul_aa](https://github.com/secretflow/spu/blob/main/libspu/mpc/semi2k/arithmetic.cc#L260) kernel, the source of these bytes is `open(x-a,y-b)`. The bytes are linear to the shape of x and y. So 4x4 has 16 times of bytes compared with 1x1.

**Comment by winnylyc** [comment]
Thank you for your quick response! It is a fantastic optimization!

I'm having difficulty comprehending the source code, and I would appreciate it if you could review my understanding of the optimization. Here's my interpretation:
Without optimization: there are 64 element-wise multiplications, requiring 64 pairs of Beaver triples. For each pair of triples, there needs an individual open operation, so there would be 64 times of bytes compared with 1x1.
With optimization: there are still 64 pairs of beaver triples, but the generated shares (e.g., [a] for x) share the same value for each element in the input matrix.
In the specific example from my test case, each element in the input matrix undergoes four multiplications. For example, x_1,1 will multiply with y_1,1/y_1,2/y_1,3/y_1,4, so there are four triples, (a_1,1,1, b_1,1,1, c_1,1,1)/ (a_1,1,2, b_1,2,1, c_1,1,2)/ (a_1,1,3, b_1,3,1, c_1,1,3)/ (a_1,1,4, b_1,4,1, c_1,1,4). With the optimization implemented, a_1,1,1=a_1,1,2=a_1,1,3=a_1,1,4. This means that all the multiplications related to x_1,1 will only require a single open operation with  (x_1,1 - a_1,1)

I have a few additional questions regarding this problem. Sorry for taking your time.

1. Does this optimization introduce additional costs in generating Beaver triples? In other words, can we consider it as an optimizattion in the online phase that incurs some extra costs during the offline phase?
2. How can I accurately benchmark the offline phase? Specifically, how can I obtain the communication cost associated with generating Beaver triples? Additionally, considering that the default setting for semi2k is TrustedFirstParty, can I assume that there is no communication cost involved in generating Beaver triples?

Thanks, truly appreciate it!

**Comment by winnylyc** [comment]
By the way, I'm curious about the source of this optimization. Is it derived from a specific research paper, or is it a novel optimization developed and implemented by the SPU team?

**Comment by llCurious** [comment]
> Thank you for your quick response! It is a fantastic optimization!
> 
> I'm having difficulty comprehending the source code, and I would appreciate it if you could review my understanding of the optimization. Here's my interpretation: Without optimization: there are 64 element-wise multiplications, requiring 64 pairs of Beaver triples. For each pair of triples, there needs an individual open operation, so there would be 64 times of bytes compared with 1x1. With optimization: there are still 64 pairs of beaver triples, but the generated shares (e.g., [a] for x) share the same value for each element in the input matrix. In the specific example from my test case, each element in the input matrix undergoes four multiplications. For example, x_1,1 will multiply with y_1,1/y_1,2/y_1,3/y_1,4, so there are four triples, (a_1,1,1, b_1,1,1, c_1,1,1)/ (a_1,1,2, b_1,2,1, c_1,1,2)/ (a_1,1,3, b_1,3,1, c_1,1,3)/ (a_1,1,4, b_1,4,1, c_1,1,4). With the optimization implemented, a_1,1,1=a_1,1,2=a_1,1,3=a_1,1,4. This means that all the multiplications related to x_1,1 will only require a single open operation with (x_1,1 - a_1,1)
> 
> I have a few additional questions regarding this problem. Sorry for taking your time.
> 
> 1. Does this optimization introduce additional costs in generating Beaver triples? In other words, can we consider it as an optimization in the online phase that incurs some extra costs during the offline phase?
> 2. How can I accurately benchmark the offline phase? Specifically, how can I obtain the communication cost associated with generating Beaver triples? Additionally, considering that the default setting for semi2k is TrustedFirstParty, can I assume that there is no communication cost involved in generating Beaver triples?
> 
> Thanks, truly appreciate it!

Hi @winnylyc . Thx for you attention in the detailed Matmul implementation.

To be brief, the [code](https://github.com/secretflow/spu/blob/main/libspu/mpc/semi2k/arithmetic.cc#L256-L257) here actually use the vectorization technique to reduce the overhead. 

Without the vectorization, it is true that a (4, 4) @ (4, 4) matmul will incur 64 standalone scalar multiplications, which should be 64x communication that of 1x1.

To multiply two shared matrices, we here take shared random matrices instead of using independent multiplication triplets (some sort of vectorization). As a consequence, we only need triplets of shape 4x4 to mask x and 4x4 to mask y, which is exactly 16x communication  that of 1x1.

Regarding your following questions:
1. Nope. The offline costs actually should be lower since we only need to generate mltiplication triplets of shape 4x4, rather than 4x4x4. This is a widely-adopted optimzation, and you can refer to Section IV-A in SecureML: A System for Scalable Privacy-Preserving Machine Learning for details.
2. Currently we do not count the offline costs. BTW your understanding is correct. If you use TFP setting, there should be no communication cost to **generate** the triples.

**Comment by winnylyc** [comment]
Thanks for your patient response. It really helps me a lot on understanding MPC primitives!

**Comment by winnylyc** [comment]
Sorry for disturbing you again. After further reading SecureML, I am curious about the implementation of another optimization described in Section IV-A, which is opening the data samples just once at the first iteration and use the opened masked data samples for other iterations. It seems to be an optimization that can be generalized to all matmul. 

Here is an example, I implement `x matmul y` and `x matmul y`. Since the value of `x` keeps unchanged, it seems to be possible to just open maksed `x` once.
```
import secretflow as sf
import spu
import jax.numpy as jnp
import numpy as np

def test_function(x, y, z):
    result1 = jnp.matmul(x, y)
    result2 = jnp.matmul(x, z)
    return result1, result2

sf.shutdown()
sf.init(['alice', 'bob'], address='local')

alice, bob = sf.PYU('alice'), sf.PYU('bob')

semi2k_config = sf.utils.testing.cluster_def(
    parties=['alice', 'bob'],
    runtime_config={
        'protocol': spu.spu_pb2.SEMI2K,
        'field': spu.spu_pb2.FM64,
        "enable_pphlo_profile": True,
        "enable_hal_profile": True,
        # "experimental_disable_vectorization": True,
    },
)

spu_device = sf.SPU(semi2k_config)

x_plain = np.array([[5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.]])
# x_plain =  np.array([5.])

y_plain = np.array([[5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.]])
# y_plain = np.array([5.])

z_plain = np.array([[5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.], [5., 5., 5., 5.]])

x = sf.to(alice, x_plain)
y = sf.to(bob, y_plain)
z = sf.to(bob, z_plain)

result1, result2 = spu_device(test_function)(x, y, z)

print(sf.reveal(result1), sf.reveal(result2))
```
Here is the profile.
```
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:191] HLO profiling: total time 0.002659319
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:194] - pphlo.dot, executed 2 times, duration 0.002659319s, send bytes 768
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:191] HAL profiling: total time 0.001791601
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:194] - f_mmul, executed 2 times, duration 0.001791601s, send bytes 768
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:191] MPC profiling: total time 0.001293149
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:194] - mmul_aa, executed 2 times, duration 0.000735804s, send bytes 512
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:194] - reshape, executed 2 times, duration 7.253e-06s, send bytes 0
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.717 [info] [api.cc:printProfilingData:194] - trunc_a, executed 2 times, duration 0.000550092s, send bytes 256
(SPURuntime(device_id=None, party=alice) pid=24331) 2024-03-12 13:17:21.718 [info] [api.cc:printProfilingData:204] Link details: total send bytes 768, send actions 4
```
According to the profile, I think the `x` is opened twice. If the optimization is feasible to this situation, the send bytes can be optimized to 512 - 128 (an open operation for 'x' can be saved).

Is there any method to call this optimization is SPU. Or will this method cause addtional cause additional cost or security issue?

Looking forward to your reply!

**Comment by winnylyc** [comment]
Thanks for your confirmation! And sorry for so many typos. 😓 I really learn a lot during this series of discussions!
